src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,397 @@
+use std::{collections::HashSet, convert::identity, sync::Arc};
+
+use common_error::DaftResult;
+
+use daft_core::schema::Schema;
+use daft_dsl::Expr;
+
+use crate::{
+    ops::{Aggregate, Project, Source},
+    LogicalPlan,
+};
+
+use super::{ApplyOrder, OptimizerRule};
+
+#[derive(Default)]
+pub struct PushDownProjection {}
+
+impl PushDownProjection {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn try_optimize_project(
+        &self,
+        projection: &Project,
+        plan: &LogicalPlan,
+    ) -> DaftResult<Option<Arc<LogicalPlan>>> {
+        let upstream_plan = projection.input.clone();
```

Comment:
Nit: Is there a reason that this clone is required? You should be able to directly take `projection.input.as_ref()` for the `match` statement and defer cloning until needed.